Mainnet/DividendToken/DividendToken/DividendToken.cs

```diff
@@ -0,0 +1,288 @@
+﻿using Stratis.SmartContracts;
+using Stratis.SmartContracts.Standards;
+
+[Deploy]
+public class DividendToken : SmartContract, IStandardToken
+{
+
+    public ulong Dividends
+    {
+        get => PersistentState.GetUInt64(nameof(this.Dividends));
+        private set => PersistentState.SetUInt64(nameof(this.Dividends), value);
+    }
+
+    private Account GetAccount(Address address) => PersistentState.GetStruct<Account>($"Account:{address}");
+
+    private void SetAccount(Address address, Account account) => PersistentState.SetStruct($"Account:{address}", account);
+
+
+    public DividendToken(ISmartContractState state, ulong totalSupply, string name, string symbol)
+        : base(state)
+    {
+        this.TotalSupply = totalSupply;
+        this.Name = name;
+        this.Symbol = symbol;
+        this.SetBalance(Message.Sender, totalSupply);
+    }
+
+    /// <summary>
+    /// It is advised that deposit amount should to be evenly divided by total supply, 
+    /// otherwise small amount of satoshi may lost(burn)
+    /// </summary>
+    public override void Receive()
+    {
+        Dividends += Message.Value;
+    }
+
+    public bool TransferTo(Address to, ulong amount)
+    {
+        UpdateAccount(Message.Sender);
+        UpdateAccount(to);
+
+        return TransferTokensTo(to, amount);
+    }
+
+    public bool TransferFrom(Address from, Address to, ulong amount)
+    {
+        UpdateAccount(from);
+        UpdateAccount(to);
+
+        return TransferTokensFrom(from, to, amount);
+    }
+
+    Account UpdateAccount(Address address)
+    {
+        var account = GetAccount(address);
+        var newDividends = GetWithdrawableDividends(address, account);
+
+        if (newDividends > 0)
+        {
+            account.DividendBalance = checked(account.DividendBalance + newDividends);
+            account.CreditedDividends = Dividends;
+            SetAccount(address, account);
+        }
+
+        return account;
+    }
+
+    private ulong GetWithdrawableDividends(Address address, Account account)
+    {
+        var newDividends = Dividends - account.CreditedDividends;
+        var notCreditedDividends = checked(GetBalance(address) * newDividends);
```

Comment:
Partial reason for using `UInt256` for token value properties right here. Huge chance of overflow exception thrown multiplying `ulong * ulong = ulong`. This can be a critical issue.

Reply:
For added context, max ulong value is `18446744073709551615`, square root of that is `4294967296` which is equal to `42.94967296` tokens. Making any combination of values higher than that multiplied by each other result in the overflow exception metioned.

Reply:
Max value of `newDividends` can be Total Supply of Cirrus coin which is 100 million and max value of `GetBalance(address)` can be Total Supply of dividend token so in this case, it can overflow if total supply is higher than 184.467.440.737 number.  It is less likely that total supply can be such big number but it  can be.  We may restrict total supply for higher than 184.467.440.737 number.

Reply:
I do not suggested pushing out a smart contract that has these limitations or possible overflow exceptions. Even if the chance is small, it will lock tokens from being able to be withdrawn and result in loss of funds.